src/base/bittorrent/sessionimpl.cpp

```diff
@@ -628,8 +633,22 @@ SessionImpl::~SessionImpl()
     m_asyncWorker->clear();
     m_asyncWorker->waitForDone();
 
-    qDebug("Deleting libtorrent session...");
+    auto *nativeSessionProxy = new lt::session_proxy(m_nativeSession->abort());
     delete m_nativeSession;
+
+    qDebug("Deleting resume data storage...");
+    delete m_resumeDataStorage;
+    LogMsg(tr("Saving resume data completed."));
+
+    auto *sessionTerminateThread = QThread::create([nativeSessionProxy]()
+    {
+        qDebug("Deleting libtorrent session...");
+        delete nativeSessionProxy;
+    });
+    connect(sessionTerminateThread, &QThread::finished, sessionTerminateThread, &QObject::deleteLater);
+    sessionTerminateThread->start();
+    if (sessionTerminateThread->wait(shutdownDeadlineTimer))
+        LogMsg(tr("BitTorrent session finished."));
```

Comment:
Finished what? Saving? And shouldn't there also be a message if session was killed or maybe only then?

Reply:
> Finished what?

The session itself is finished. (My English is bad.)
How to say it better?

Reply:
> And shouldn't there also be a message if session was killed or maybe only then?

I have not found a reliable (and not too complicated) way to capture this moment. Therefore, the absence of "session finished" message in the log can be considered an indicator that the session is most likely interrupted.

Reply:
It would be better if the message reflected when a session ends early rather than as usual (successful).

Reply:
>The session itself is finished. (My English is bad.)
How to say it better?

"BitTorrent session ended cleanly"? I don't know, can't think of something good.

Reply:
> > The session itself is finished. (My English is bad.)
> > How to say it better?
> 
> "BitTorrent session ended cleanly"? I don't know, can't think of something good.

Maybe "BitTorrent session is stopped"? Or just leave it as it is?

It's just an event log. Let's not puzzle ourselves. The main thing is to make it clear what we are talking about when we diagnose a problem.

Reply:
"BitTorrent session ended successfully"? Perhaps I should add that if you don't see this message, it means the session ended early? Or leave it as it is.

Reply:
>Maybe "BitTorrent session is stopped"?

Doesn't seem better.
>The main thing is to make it clear what we are talking about when we diagnose a problem.

Exactly. I had to go read QThread's wait() docs to see what's happening here because "BitTorrent session finished." nor "is stopped" disambiguate if session was killed or exited normally.

Reply:
> "BitTorrent session finished." nor "is stopped" disambiguate if session was killed or exited normally.

Strange... The difference between "finished" and "interrupted", "killed" etc. is obvious to me.
Would "successfully finished" be more clear?

Reply:
> I had to go read QThread's wait() docs to see what's happening here

In case it seems to someone that it could be done as the following:
```c++
if (wait())
    log("succeeded");
else
    log("failed");
```
We do not interrupt the thread at this particular place, we just stop waiting for it to finish and continue to terminate the application. It still has a chance to finish normally, we just don't care about it anymore.

Reply:
> > I had to go read QThread's wait() docs to see what's happening here
> 
> In case it seems to someone that it could be done as the following:
> 
> ```c++
> if (wait())
>     log("succeeded");
> else
>     log("failed");
> ```
> 
> We do not interrupt the thread at this particular place, we just stop waiting for it to finish and continue to terminate the application. It still has a chance to finish normally, we just don't care about it anymore.

Or maybe it should be like the following (I'm not sure it is really needed):
```
if (wait())
    log("Session successfully finished.");
else
    log("Session shutdown timed out.");
```

Reply:
>Would "successfully finished" be more clear?

Of course.

Reply:
> > Would "successfully finished" be more clear?
> 
> Of course.

It is already done.
The last question is https://github.com/qbittorrent/qBittorrent/pull/20797#discussion_r1589956591

Reply:
>The last question is https://github.com/qbittorrent/qBittorrent/pull/20797#discussion_r1589956591

I would say yes. How else would someone search for it in their logs. But if you can do this i now don't see the reason for the "Succeeded" message... i mean that's the expected behavior.

Reply:
> But if you can do this i now don't see the reason for the "Succeeded" message... i mean that's the expected behavior.

The reason is to detect a premature shutdown caused by external causes, which cannot be detected from within the application, so that the absence of this message will be its indicator.

Reply:
>premature shutdown caused by external causes

Damn, all this time i was talking about "kill" of the session by the setting in this PR... Anyway i say include the `else` part, if you want.

Reply:
> Anyway i say include the `else` part, if you want.

Done.